base/ltfinal.dtx

```diff
@@ -242,7 +242,70 @@
 % \end{macro}
 % \end{macro}
 %
-%
+% \begin{macro}{trace_stack_levels}
+%   Now define the Lua function to emulate \cs{tracingstacklevels} and
+%   install it in the \texttt{input_level_string} callback.
+% \changes{v2.0k}{2016/01/05}{Remove duplicated code}
+%    \begin{macrocode}
+%</2ekernel>
+%<*2ekernel|latexrelease>
+%<latexrelease>\IncludeInRelease{2021/06/01}{trace_stack_levels}%
+%<latexrelease>                 {Lua trace_stack_levels function}%
+\ifx\directlua\@undefined
+\else
+%<*2ekernel>
+  \expanded{%
+    \everyjob{\the\everyjob
+    \noexpand%\directlua
+%</2ekernel>
+    \directlua{%
+      local function trace_stack_levels (input_ptr)
+        local tracingstacklevels = tex.count.tracingstacklevels
+        if tex.tracingmacros > 0 or input_ptr < tracingstacklevels then
+          if tracingstacklevels > 0 then
+            if input_ptr < tracingstacklevels then
+              return "\string\n\string~" .. string.rep(".",input_ptr)
+            else
+              return "\string~\string~"
+            end
+          else
+            return "\string\n"
+          end
+        else
+          return ""
+        end
+      end
+%<*latexrelease>
+      if luatexbase.callbacktypes['input_level_string'] and %
```

Comment:
This shouldn't hurt but since `ltluatex.lua` doesn't rollback the test shouldn't be necessary.

Reply:
My reasoning was that in case of rollforward someone might be using an `ltluatex` without that callback installed, so the test in the `latexrelease` version prevents adding to an undefined callback.  Right or did I miss something?

Reply:
I thought that when rolling forward, `ltluatex` is rolled forward too, but that probably depends on the user (if they install the `ltluatex.lua`or not). So keeping the test is probably safer.